### PR TITLE
[MRG+2] ENH: padding kwarg for psd_welch

### LINKED ISSUE #3621: n_fft issue
Changing n_fft does not change the frequency bins when using
`mne.time_frequency.psd_welch`

### PR DESCRIPTION
**Pull request by mmagnuski**
Fixes #3621 
by adding `padding` kwarg to `psd_welch`.
I'll add an entry to whats_new once comments are resolved and CIs green.

### CODE REVIEW

File: mne/time_frequency/psd.py [hunk 1]
```diff
@@ -11,17 +11,19 @@
 from .multitaper import _psd_multitaper
 
 
-def _psd_func(epoch, noverlap, nfft, fs, freq_mask, func):
+def _psd_func(epoch, noverlap, nfft, fs, freq_mask, func, padding):
     """Aux function."""
     return func(epoch, fs=fs, nperseg=nfft, noverlap=noverlap,
-                nfft=nfft, window='hann')[2][..., freq_mask, :]
+                nfft=padding, window='hann')[2][..., freq_mask, :]
```
Comment (agramfort) [inline mne/time_frequency/psd.py]: I don't like this. nfft and padding have different meanings.

Comment (mmagnuski) [inline mne/time_frequency/psd.py]: This is the scipy.spectral.spectrogram API. It differs from mne's psd_welch: `nperseg` is the length of each welch window and `nfft` is the final length of each window. So if nfft > nperseg it results in zero padding nperseg segments to nfft. psd_welch already uses nfft to mean nperseg, so I added padding (or `pad_to` or a different kwarg name if you prefer). We could also change psd_welch a little to use the same logic as scipy spectrogram (nperseg and nfft) but that would require deprecation (although the behavior could stay backward compatible - the meaning of the parameters would change).

Comment (dengemann) [inline mne/time_frequency/psd.py]: @mmagnuski can you summarize the new logic? You add a padding parameter which can override the n_fft if it is not None?

Comment (mmagnuski) [inline mne/time_frequency/psd.py]: Previously both nperseg and nfft were populated with psd_welch nfft - which means that padding was never used. Now there is a separate kwarg for padding that is used only when not None and > nfft. I agree that naming is unfortunate - ideally nfft should control padding if > nperseg (just like in scipy). We can have such behavior in mne but this would probably require a deprecation...

**Comment by mmagnuski** [comment]
Or not - depending on the wording of the docs and your opinion. :)
We can pass nfft to scipys nfft and add nperseg kwarg.

**Comment by dengemann** [comment]
@mmagnuski I'm trying to better understand a) the motivation for this b) how it changes the exisiting behavior with regard to `n_fft`. Maybe you want to share your insights :)

**Comment by mmagnuski** [comment]
Sure, the motivation for this is adding control over padding - see #3621. It wasn't a long discussion but the general agreement was to add control over padding, likely through additional kwarg. Currently you can't have padding - only segment length is controlled. 
Segment length is termed `n_fft` in mne and `nperseg` in scipy. Mne naming seems to stem from the fact that there was no ability to have zero-padding, so segment length was the same as n_fft. 

Introducing control over padding requires either:
* adding `nperseg` kwarg that defines segment length, and make `n_fft` control padding when n_fft > nperseg. But because previously `n_fft` was closer in meaning to nperseg it was adjusted down to `len(inst.times)` when `n_fft > len(inst.times)`. So this kind of change would rather require deprecation.
* another option is to add `padding` or `pad_to`, as done in this PR. A separate kwarg that controls padding, while `n_fft` controls segment length. Padding takes place only when padding > n_fft. The downside here is that `n_fft` works as `nperseg` which may be conterintuitive.

I also cover the logic proposed here in my previous comments - in case you didn't see all of them.
And I actually would prefer adding `nperseg` instead of `padding` - but this requires some kind of deprecation (n_fft >signal_length would do padding).

**Comment by Eric89GXL** [comment]
I'm okay with deprecation here if it gets equivalent API with SciPy.

**Comment by agramfort** [comment]
what disturb me is that nfft = len(signal) + padding
it was a semantic mismatch.

Why would you want to do some padding in the first place? Have non power of
2 segments but fast ffts?

**Comment by mmagnuski** [comment]
@agramfort
Yeah, I understand the problem with semantic mismatch. What about solution 2?

It is not so much my need to add padding - but in #3621 you agreed it should be added (at least I read your "I agree" this way :) ). In general non power of 2 is one case but interpolation is another. I remember I was a bit surprised when I started using mne and found out that you can't  set window length and padding - just as the person who submitted

**Comment by mmagnuski** [comment]
sorry, I'm on my phone, pushed wrong button by accident.
...the person who submitted #3621.

**Comment by agramfort** [comment]
please show me how high level API would look like

**Comment by mmagnuski** [comment]
with solution 2 that would be: 
```python
psds, freqs = psd_welch(raw, nperseg=128, n_fft=256)
```

with solution 1: 
```python
psds, freqs = psd_welch(raw, n_fft=128, pad_to=256)
```

**Comment by agramfort** [comment]
ok fine with solution 2 with nperseg defaulting to None meaning use n_fft
so we have no deprecation.

**Comment by mmagnuski** [comment]
n_fft > signal length would do zero-padding, while previously it would be set back to signal length. So there should be some deprecation info if nperseg=None and n_fft > signal length, right? Or do you suggest that n_fft >

**Comment by mmagnuski** [comment]
(eh, same problem, sorry)
n_fft > signal length should not do zero padding when nperseg=None?

**Comment by agramfort** [comment]
well calling welch on a signal that is less than nperseg is not really
doing welch
and should probably be not done. You should use mulit-tapers for such short
signals

**Comment by Eric89GXL** [comment]
> well calling welch on a signal that is less than nperseg is not really doing welch and should probably be not done.

Agreed. I think it's okay to raise an error if someone tries to do this. If it breaks some existing code and forces someone who really wants to do it to now be explicit about it, I think that's okay

**Comment by mmagnuski** [comment]
Ok, so if nperseg=None and n_fft > signal_length -> raise an error stating that if one really wants this nperseg has to specified.

**Comment by Eric89GXL** [comment]
Sounds reasonable to me. This allows us to proceed with solution 2 (SciPy API), right?

**Comment by agramfort** [comment]
yes

**Comment by mmagnuski** [comment]
I've added `nperseg` kwarg, but everywhere in mne there is rather `n_something` than `nsomething`. For example - `psd_welch` has `n_fft`, not `nfft`. So `n_per_seg` might be more consistent with the rest of naming (although I don't like two underscores there). I've also changed the docstring a little, so take a look if you like the changes.

**Comment by mmagnuski** [comment]
Just in case another option could be `segment_len` (or maybe `n_segment`) or `window_len` (etc.)

**Comment by Eric89GXL** [comment]
I think it's better to stay close to SciPy. `n_per_seg` seems like the right compromise between equivalence with SciPy (which we can't get fully anyway anymore) and our internal style practices

**Comment by mmagnuski** [comment]
Ok, I'll change it to `n_per_seg` then.

**Comment by Eric89GXL** [comment]
Ready for review/merge from your end?

**Comment by mmagnuski** [comment]
Yes, it should be good for review/merge now.

File: mne/time_frequency/psd.py [hunk 2]
```diff
@@ -89,7 +96,11 @@ def _psd_welch(x, sfreq, fmin=0, fmax=np.inf, n_fft=256, n_overlap=0,
     x = x.reshape(-1, n_times)
 
     # Prep the PSD
-    n_fft, n_overlap = _check_nfft(n_times, n_fft, n_overlap)
+    if n_per_seg is None and n_fft > n_times:
+        raise ValueError('If n_per_seg is None n_fft is not allowed to be >'
+                         ' n_times. If you want zero-padding, you have to set'
+                         ' n_per_seg to relevant length.')
```
Comment (agramfort) [inline mne/time_frequency/psd.py]: say what value of n_times and were given compared to n_fft

**Comment by agramfort** [comment]
besides LGTM

File: mne/time_frequency/psd.py [hunk 3]
```diff
@@ -11,17 +11,18 @@
...
-def _check_nfft(n, n_fft, n_overlap):
-    """Helper to make sure n_fft and n_overlap make sense."""
-    n_fft = n if n_fft > n else n_fft
-    n_overlap = n_fft - 1 if n_overlap >= n_fft else n_overlap
-    return n_fft, n_overlap
+def _check_nfft(n, n_fft, n_per_seg, n_overlap):
+    """Helper to make sure n_fft, n_per_seg and n_overlap make sense."""
+    n_per_seg = n_fft if n_per_seg is None or n_per_seg > n_fft else n_per_seg
+    n_per_seg = n if n_per_seg > n else n_per_seg
+    n_overlap = n_per_seg - 1 if n_overlap >= n_per_seg else n_overlap
```
Comment (Eric89GXL) [inline mne/time_frequency/psd.py]: Maybe we covered this already, but this should probably just be an error, no?

File: mne/time_frequency/psd.py [hunk 4]
```diff
@@ -65,13 +66,19 @@ def _psd_welch(x, sfreq, fmin=0, fmax=np.inf, n_fft=256, n_overlap=0,
     fmax : float
         The upper frequency of interest.
     n_fft : int
-        The length of the tapers ie. the windows. The smaller
-        it is the smoother are the PSDs. The default value is 256.
-        If ``n_fft > len(inst.times)``, it will be adjusted down to
-        ``len(inst.times)``.
+        The length of FFT used. If n_per_seg is None n_fft sets the length of the
```
Comment (Eric89GXL) [inline mne/time_frequency/psd.py]: This is good enough:
```
The length of FFT used, must be ``>= n_per_seg`` (default: 256).
The segments will be zero-padded if ``n_fft > n_per_seg``.
```

File: mne/time_frequency/psd.py [hunk 5]
```diff
@@ -65,13 +66,19 @@ def _psd_welch(x, sfreq, fmin=0, fmax=np.inf, n_fft=256, n_overlap=0,
...
+        If ``n_fft > n_times`` and n_per_seg is None an error is thrown.
+        The default value is 256.
     n_overlap : int
         The number of points of overlap between blocks. Will be adjusted
-        to be <= n_fft. The default value is 0.
+        to be <= n_per_seg. The default value is 0.
+    n_per_seg : int | None
+        Length of each Welch segment. The smaller it is with respect to the
+        signal length the smoother are the PSDs. Defaults to None, which sets
+        n_per_seg equal to n_fft. If n_per_seg is smaller than n_fft, each window
```
Comment (Eric89GXL) [inline mne/time_frequency/psd.py]: If you add the bits about padding above, they don't need to be here.

File: mne/time_frequency/psd.py (hunk 2 above)
Comment (Eric89GXL) [inline mne/time_frequency/psd.py]: this check should go in `_check_nfft`

File: mne/time_frequency/psd.py (hunk 3 above)
Comment (mmagnuski) [inline mne/time_frequency/psd.py]: that's the previous behavior, I can change it to an error if you prefer.

File: mne/time_frequency/psd.py (hunk 4 above)
Comment (mmagnuski) [inline mne/time_frequency/psd.py]: Yes, that's better, thanks!

File: mne/time_frequency/psd.py (hunk 5 above)
Comment (mmagnuski) [inline mne/time_frequency/psd.py]: ok

File: mne/time_frequency/psd.py (hunk 3 above)
Comment (Eric89GXL) [inline mne/time_frequency/psd.py]: I'd rather have people be explicit if they want this behavior. This looks like it's silent-errant-behavior prone to me

File: mne/time_frequency/psd.py (hunk 2 above)
Comment (mmagnuski) [inline mne/time_frequency/psd.py]: ok, I'll move it there

**Comment by mmagnuski** [comment]
Codecov is confused, it says "Absolute coverage decreased by -12.11% ..." and lists many tests as not tested...

**Comment by mmagnuski** [comment]
and Travis did not run

**Comment by Eric89GXL** [comment]
@agramfort +1 for merge from you? looks like comments have been addressed

@mmagnuski sometimes there is a delay, everything looks good now. +1 for merge from me

File: mne/time_frequency/psd.py [hunk 6]
```diff
@@ -11,17 +11,26 @@
...
+    if n_per_seg is None and n_fft > n:
+        msg = ('If n_per_seg is None n_fft is not allowed to be > n_times. If '
+               'you want zero-padding, you have to set n_per_seg to relevant '
+               'length. Got n_fft of {} while signal length is {}.')
+        raise ValueError(msg.format(n_fft, n))
+    n_per_seg = n_fft if n_per_seg is None or n_per_seg > n_fft else n_per_seg
+    n_per_seg = n if n_per_seg > n else n_per_seg
+    if n_overlap >= n_per_seg:
+        msg = ('n_overlap cannot be greater than n_per_seg (or n_fft). Got '
+               'n_overlap of {} while n_per_seg is {}.')
```
Comment (Eric89GXL) [inline mne/time_frequency/psd.py]: FYI for future reference by convention we 1. don't split these across multiple vars/commands and 2. use/prefer `%` (even though there is talk of deprecation)

Comment (mmagnuski) [inline mne/time_frequency/psd.py]: Thanks for stating what the preference is - I think I saw this using format and a separate var with text) a few times in Jona's code so I thought it is ok. I'll remember now. :) I can change it here too.

Comment (Eric89GXL) [inline mne/time_frequency/psd.py]: Sometimes `.format` is a lot more convenient (e.g., `mne/datasets/utils.py` I think?), and we do use it in those cases. But it's not a big deal, so don't bother unless someone else has changes they want you to make

**Comment by agramfort** [comment]
please update what's new and then +1 for MRG

**Comment by agramfort** [comment]
yes ok for me

**Comment by mmagnuski** [comment]
@Eric89GXL the errors we added make a test fail:
```
======================================================================
ERROR: Test creating raw from array.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/numpy/testing/decorators.py", line 147, in skipper_func
    return f(*args, **kwargs)
  File "/home/travis/build/mne-tools/mne-python/mne/io/array/tests/test_array.py", line 95, in test_array_raw
    raw2.plot_psd(tmax=np.inf, average=True, spatial_colors=False)
  File "<string>", line 2, in plot_psd
  File "/home/travis/build/mne-tools/mne-python/mne/utils.py", line 709, in verbose
    return function(*args, **kwargs)
  File "/home/travis/build/mne-tools/mne-python/mne/io/base.py", line 1638, in plot_psd
    spatial_colors=spatial_colors)
  File "<string>", line 2, in plot_raw_psd
  File "/home/travis/build/mne-tools/mne-python/mne/utils.py", line 709, in verbose
    return function(*args, **kwargs)
  File "/home/travis/build/mne-tools/mne-python/mne/viz/raw.py", line 598, in plot_raw_psd
    n_overlap=n_overlap, n_jobs=n_jobs)
  File "<string>", line 2, in psd_welch
  File "/home/travis/build/mne-tools/mne-python/mne/utils.py", line 709, in verbose
    return function(*args, **kwargs)
  File "/home/travis/build/mne-tools/mne-python/mne/time_frequency/psd.py", line 200, in psd_welch
    n_jobs=n_jobs, verbose=verbose)
  File "<string>", line 2, in psd_array_welch
  File "/home/travis/build/mne-tools/mne-python/mne/utils.py", line 709, in verbose
    return function(*args, **kwargs)
  File "/home/travis/build/mne-tools/mne-python/mne/time_frequency/psd.py", line 114, in psd_array_welch
    n_overlap)
  File "/home/travis/build/mne-tools/mne-python/mne/time_frequency/psd.py", line 26, in _check_nfft
    ' signal length is %d.') % (n_fft, n))
ValueError: If n_per_seg is None n_fft is not allowed to be > n_times. If you want zero-padding, you have to set n_per_seg to relevant length. Got n_fft of 2048 while signal length is 1803.
```
Previous behavior was to set n_fft equal to signal length if n_fft > signal length so I'll change this test by setting n_fft to signal length.

**Comment by mmagnuski** [comment]
All builds are green except one:
```
======================================================================
FAIL: Test object size estimation.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/envs/test/lib/python3.5/site-packages/mne-0.14.dev0-py3.5.egg/mne/tests/test_utils.py", line 111, in test_object_size
    msg='%s < %s < %s:\n%s' % (lower, size, upper, obj))
AssertionError: False is not true : 0 < 68 < 60:
foo
    'False is not true : 0 < 68 < 60:\nfoo' = self._formatMessage('False is not true : 0 < 68 < 60:\nfoo', "%s is not true" % safe_repr(False))
>>  raise self.failureException('False is not true : 0 < 68 < 60:\nfoo')
```

**Comment by Eric89GXL** [comment]
Yep, failure is unrelated. Thanks @mmagnuski !

**Comment by mmagnuski** [comment]
Thanks @Eric89GXL @agramfort 🚀

**Comment by Eric89GXL** [comment]
Onto #4000 :)
